apps/studio/src/server/modules/resource/resource.service.ts

```diff
@@ -441,25 +450,147 @@ export const getResourceFullPermalink = async (
   return `/${permalinkTree.join("/")}`
 }
 
-export const publishResource = async (
+export const publishPageResource = async (
   logger: Logger<string>,
   siteId: number,
   resourceId: string,
   userId: string,
 ) => {
   // Step 1: Create a new version
-  const addedVersionResult = await incrementVersion({
-    siteId,
-    resourceId,
-    userId,
-  })
+  const by = await db
+    .selectFrom("User")
+    .selectAll()
+    .where("id", "=", userId)
+    .executeTakeFirstOrThrow(
+      () =>
+        new TRPCError({
+          code: "PRECONDITION_FAILED",
+          message: "Please ensure that you have logged in",
+        }),
+    )
+
+  const addedVersionResult = await db
+    .transaction()
+    .setIsolationLevel("serializable")
+    .execute(async (tx) => {
+      const fullResource = await getFullPageById(tx, {
+        resourceId: Number(resourceId),
+        siteId,
+      })
+
+      if (!fullResource) {
+        throw new TRPCError({
+          code: "PRECONDITION_FAILED",
+          message:
+            "Please ensure you are attempting to publish a page that exists",
+        })
+      }
+
+      const version = await incrementVersion({
+        tx,
+        siteId,
+        resourceId,
+        userId,
+      })
+
+      const previousVersion = await tx
+        .selectFrom("Version")
+        .where("Version.versionNum", "=", Number(version.versionId) - 1)
+        .where("Version.resourceId", "=", resourceId)
+        .select("Version.id")
+        .executeTakeFirst()
+
+      await logPublishEvent(tx, {
+        by,
+        delta: {
+          before: previousVersion?.id
+            ? { versionId: previousVersion.id }
+            : null,
+          after: version,
+        },
+        eventType: "Publish",
+        metadata: fullResource,
+      })
+
+      return version
+    })
 
   // Step 2: Trigger a publish of the site
   await publishSite(logger, siteId)
 
   return addedVersionResult
 }
 
+// NOTE: The distinction here between `publishResource` and `publishPageResource` is that
+// this should be used for publishes that do not incur a change to `Blob.content`
+// and hence, don't incur a log to the `Version` table
+export const publishResource = async (
+  by: string,
```

Comment:
nit - should we use the type from User["id"] instead of just a string?

Reply:
i think `User["id"]` is a `string` (iirc), but open to making this change